typing/typedecl.ml

```diff
@@ -226,6 +253,34 @@ let make_constructor env type_path type_params sargs sret_type =
       widen z;
       targs, Some tret_type, args, Some ret_type
 
+(* Check that the argument to a GADT constructor is compatible with unboxing
+   the type, given the existential variables introduced by this constructor. *)
+let rec check_unboxed_gadt_arg loc ex env ty =
+  match get_unboxed_type_representation env ty with
+  | Some {desc = Tvar _ as d; _} ->
+    let f t = (Btype.repr t).desc = d in
+    if List.exists f ex then
+      raise(Error(loc, Wrong_unboxed_type
+                    "it might contain both float and non-float values"))
+  | Some {desc = Tarrow _ | Ttuple _ | Tpackage _ | Tobject _ | Tnil
+                 | Tvariant _; _} ->
+    ()
+    (* A comment in [Translcore.transl_exp0] claims the above cannot be
+       represented by floats. *)
+  | Some {desc = Tconstr (p, args, _); _} ->
+    let tydecl = Env.find_type p env in
+    assert (not (Builtin_attributes.has_unboxed tydecl.type_attributes));
+    if tydecl.type_kind = Type_abstract then
+      List.iter (check_unboxed_gadt_arg loc ex env) args
+  | Some {desc = Tfield _ | Tlink _ | Tsubst _; _} -> assert false
+      (* TODO: check with Jacques that Tfield is indeed impossible *)
+  | Some {desc = Tunivar _; _} -> ()
```

Comment:
The scope of a Tunivar can only be a type expression. So for instance it cannot include multiple branches of a datatype definition (or multiple fields of a record), but it can include a whole polymorphic variant or object type expression.
Also concerning the above Tfield, it can only appear under a Tobject, and since you do not recurse there, you never meet one.